tests/README.md

```diff
@@ -0,0 +1,178 @@
+# Crypto tests
+
+## Running crypto Perl tests
+
+Crypto tests require the Math::GMP Perl library, make sure it is installed on you system before running the tests.
+
+Installing dependencies (using cpan):
+
+```
+cpan
+cpan> install Math::BigInt::GMP
+cpan> install Digest::Keccak
+```
+
+Running tests:
+
+```
+TESTPATH=/path/to/monero/tests
+cd $TESTPATH
+perl -I $TESTPATH cryptotest.pl
+```
+
+Important: must include test path for perl to import cryptolib.pl
+
+## Writing new crypto tests
+
+[TODO]
+
+# Core tests
+
+## Running core tests
+
+Monero uses the Google C++ Testing Framework (`gtest`) to write unit, integration and functional tests for core and other features of the project.
+`gtest` runs on top of cmake, and you can run all tests by:
+
+```
+cd /path/to/monero
+make [-jn] debug-test # where n is number of compiler processes
+```
+
+To test a release build, replace `debug-test` with `release-test` in the previous command.
+
+One can also run individual test suites by building monero, then running `ctest` in test suite folders.
+
+Run only the hash tests:
+
+```
+cd /path/to/monero
+make [-j#] debug
+cd build/debug/tests/hash
+ctest
+```
+
+To run the same tests on a release build, replace `debug` with `release` in previous commands.
+
+## Writing new tests
+
+Based on local tests and Google's guide on creating [simple tests with gtest](https://github.com/google/googletest/blob/master/googletest/docs/Primer.md#simple-tests)
+
+Tests consist of a test harness (defined with the TEST() macro), and the test body consisting of gtest assertions.
+
+Example of a test harness:
+
+```
+TEST(test_case_name, test_name) {
+ ... test body ...
+
+}
+```
+
+As an example in Monero's [crypto unit test](./unit_tests/crypto.cpp):
+
+```
+TEST(Crypto, Ostream)
+{
+  EXPECT_TRUE(is_formatted<crypto::hash8>());
+  EXPECT_TRUE(is_formatted<crypto::hash>());
+  EXPECT_TRUE(is_formatted<crypto::public_key>());
+  EXPECT_TRUE(is_formatted<crypto::secret_key>());
+  EXPECT_TRUE(is_formatted<crypto::signature>());
+  EXPECT_TRUE(is_formatted<crypto::key_derivation>());
+  EXPECT_TRUE(is_formatted<crypto::key_image>());
+}
+
+```
+
+The assertions inside the test harness are a bit complex, but fairly straightforward.
+
+- `is_formatted<T>()` is a polymorphic function that accepts the various types of structs defined in [crypto/hash.h](../src/crypto/hash.h).
```

Comment:
If you feel the need to explain a particular test in a README, then maybe use simpler test contents since you're not documenting a particular complex test, just how to make some.

Reply:
I get your point, wasn't sure myself if this was the right place to include a walk-through. Is there a common location for walk-throughs/tutorials? The test is definitely simple. That entry was more about explaining where to edit test source files, and how to run individual tests (mb that should be more explicit?)

Reply:
If you want to add an example, just use a trivial one, since the contents don't matter. ASSERT_TRUE(1==1) :) Adding one's trivial anyway once you see it.